File: Sorting/quickSort.js

function pivotFunction(arr, start = 0, end = arr.length - 1) {
    pivot = arr[start]
    swapId = start

    for (i = start + 1; i < arr.length; i++) {
        if (pivot > arr[i]) {
            swapId++
            let temp = arr[i]
            arr[i] = arr[swapId]
            arr[swapId] = temp
        }

    }
    let temp = arr[start]
    arr[start] = arr[swapId]
    arr[swapId] = temp

    return swapId

}

function quickSort(arr, left = 0, right = arr.length - 1) {
    if (left < right) {
        let pivotIndex = pivotFunction(arr, left, right)
        quickSort(arr, left, pivotIndex - 1)
        quickSort(arr, pivotIndex + 1, right)
    }
    return arr
}
arr = [4, 1, 3, 2, 8]
console.log(quickSort(arr))

File: python/Sorting/quick-sort.py
def quickSortHelper(arr,start,end):
    pivot = arr[start]
    swapIdx = start 

    for i in range(start+1,len(arr)):
        if pivot > arr[i]:
            swapIdx+=1
            arr[swapIdx],arr[i]=arr[i],arr[swapIdx]

    arr[start],arr[swapIdx]=arr[swapIdx],arr[start]
    return swapIdx

def quickSort(arr,left,right):

    if left<right:
        pivotIdx = quickSortHelper(arr,left,right)
        quickSort(arr,left,pivotIdx-1)
        quickSort(arr,pivotIdx+1,right)
    
    return arr

arr = [100,-3,-4,4,8,2,1,5,7,6,3]
print(quickSort(arr,0,len(arr)-1))